addons/stock/product.py

```diff
@@ -84,7 +85,15 @@ def _get_domain_locations(self, cr, uid, ids, context=None):
                 location_ids = context['location']
         else:
             if context.get('warehouse', False):
-                wids = [context['warehouse']]
+                if type(context['warehouse']) == type(1):
+                    wids = [context['warehouse']]
+                elif type(context['warehouse']) in (type(''), type(u'')):
+                    domain = [('name', 'ilike', context['warehouse'])]
+                    if context.get('force_company', False):
+                        domain += [('company_id', '=', context['force_company'])]
+                    wids = warehouse_obj.search(cr, uid, domain, context=context)
+                else:
+                    wids = context['warehouse']
```

Comment:
The checking of type integer or string is necessary. The else for lists can be deleted, but does not change much the code.
The force_company checking it could also be removed